src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs

```diff
@@ -307,6 +323,21 @@ public static TimeSpan FromHours(double value)
             return Interval(value, TicksPerHour);
         }
 
+        public static TimeSpan FromHours(int hours)
+        {
+            return new TimeSpan((long)hours * TicksPerHour);
+        }
+
+        public static TimeSpan FromHours(int hours, long minutes = 0, long seconds = 0, long milliseconds = 0, long microseconds = 0)
```

Comment:
>TimeSpan [](http://example.com/codeflow?start=22&length=8)

all introduced method need to ensure no overflow and and not exceeding the max allowed value.

Reply:
Waiting for your reply in the review up there

Reply:
replied. We can try to optimize the checking as much as we can but have to ensure we produce the correct behavior.

Reply:
enable `checked` still have some perf cost too and we'll need to decide if we need to catch the thrown overflow exception and convert it to something like ThrowHelper.ThrowArgumentOutOfRange_TimeSpanTooLong(). @bartonjs do you know if we have any design guidelines around using `checked` in the APIs? I did some search but couldn't find any.

Reply:
@tannergooding  do you see any trick to efficiently detect overflow for this particular calculation? Something better than doing each one in turn and checking whether the sign changed after each multiplication, or doing it all in an int128.

Reply:
It really depends on how complicated we want to make it...

The simplest implementation is probably this:
```csharp
public static TimeSpan FromHours(int hours, long minutes = 0, long seconds = 0, long milliseconds = 0, long microseconds = 0)
{
    return FromHours(hours)
         + FromMinutes(minutes)
         + FromSeconds(seconds)
         + FromMilliseconds(milliseconds)
         + FromMicroseconds(microseconds);
}
```

This is no worse than the naive implementation a user might do and for all constant inputs, the JIT can produce a constant. It currently doesn't because the implementation of `operator +` is "bad" in several ways, but it does when that gets fixed. -- The static operator defers to an instance method and directly throws. It should be the inverse where the instance method defers to the static operator and a throw helper is used instead.

It's not, strictly speaking, the most efficient implementation for non-constant inputs. But that's probably fine at the end of the day and helps what is probably the overall more common scenario.

Reply:
Likewise, if we didn't require the `OverflowException` to exactly contain the message `"TimeSpan overflowed because the duration is too long"`, we could simply make `add` do `return new TimeSpan(checked(xTicks + yTicks))`. There would be some JIT work to still allow constant folding when `x op y` can be statically determined to not overflow still, but that seems like something generally worthwhile to do so that using `checked(...)` doesn't completely pessimize codegen and can be preferred over manual checks.

Reply:
alternatively we could also update the JIT to recognize the `if ((xTicks >> 63 == yTicks >> 63) && (xTicks >> 63 != result >> 63))` pattern and have it emit `jo` (jump if overflow) rather than the two comparisons and jumps it currently does. This would also allow the same manual checks to be more efficient.

both of those two JIT improvements seem like relatively worthwhile to help improve throughput for our own code when we need to handle such cases.

Reply:
> alternatively we could also update the JIT to recognize the `if ((xTicks >> 63 == yTicks >> 63) && (xTicks >> 63 != result >> 63))` pattern and have it emit `jo` (jump if overflow) rather than the two comparisons and jumps it currently does. This would also allow the same manual checks to be more efficient.
> 
> both of those two JIT improvements seem like relatively worthwhile to help improve throughput for our own code when we need to handle such cases.

Benchmark tests were conducted for several proposed solutions and others that came to mind. The code snippets can be found here:
https://gist.github.com/tcortega/f518f899b3b80cb62bdc3b5e8514f3bc

The outcome of the benchmarks, specifically in scenarios where an exception is always thrown, is shown in milliseconds:
```
| Method            | Platform | Mean      | Error     | StdDev    | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------ |--------- |----------:|----------:|----------:|------:|--------:|-------:|----------:|------------:|
| ManualChecks      | X64      |  4.577 us | 2.3261 us | 0.1275 us |  1.00 |    0.00 | 0.0610 |     576 B |        1.00 |
| Int128Checks      | X64      |  4.583 us | 1.1302 us | 0.0619 us |  1.00 |    0.04 | 0.0610 |     576 B |        1.00 |
| SmartInt128Checks | X64      |  5.176 us | 0.0715 us | 0.0039 us |  1.13 |    0.03 | 0.0610 |     576 B |        1.00 |
| CheckedKeyword    | X64      | 10.475 us | 2.2673 us | 0.1243 us |  2.29 |    0.09 | 0.0610 |     584 B |        1.01 |
|                   |          |           |           |           |       |         |        |           |             |
| ManualChecks      | X86      |  5.874 us | 0.1566 us | 0.0086 us |  1.00 |    0.00 | 0.0534 |     292 B |        1.00 |
| Int128Checks      | X86      |  6.027 us | 0.3815 us | 0.0209 us |  1.03 |    0.00 | 0.0534 |     292 B |        1.00 |
| SmartInt128Checks | X86      |  6.271 us | 0.6882 us | 0.0377 us |  1.07 |    0.01 | 0.0534 |     292 B |        1.00 |
| CheckedKeyword    | X86      | 15.735 us | 8.1722 us | 0.4479 us |  2.68 |    0.07 | 0.0305 |     300 B |        1.03 |
```
Results for scenarios without exceptions, simulating expected typical user inputs for the API:
```
| Method            | Platform | Mean      | Error     | StdDev    | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------ |--------- |----------:|----------:|----------:|------:|--------:|----------:|------------:|
| ManualChecks      | X64      |  4.239 ns | 4.3787 ns | 0.2400 ns |  1.00 |    0.00 |         - |          NA |
| Int128Checks      | X64      |  5.655 ns | 1.0012 ns | 0.0549 ns |  1.34 |    0.08 |         - |          NA |
| SmartInt128Checks | X64      | 10.133 ns | 0.4728 ns | 0.0259 ns |  2.40 |    0.14 |         - |          NA |
| CheckedKeyword    | X64      |  2.056 ns | 0.4019 ns | 0.0220 ns |  0.49 |    0.03 |         - |          NA |
|                   |          |           |           |           |       |         |           |             |
| ManualChecks      | X86      | 16.395 ns | 1.6665 ns | 0.0913 ns |  1.00 |    0.00 |         - |          NA |
| Int128Checks      | X86      | 52.030 ns | 7.5380 ns | 0.4132 ns |  3.17 |    0.04 |         - |          NA |
| SmartInt128Checks | X86      | 42.805 ns | 4.0426 ns | 0.2216 ns |  2.61 |    0.02 |         - |          NA |
| CheckedKeyword    | X86      | 19.274 ns | 2.3722 ns | 0.1300 ns |  1.18 |    0.01 |         - |          NA |
```
Based on these results, the ManualChecks approach seems promising. Despite being slightly less legible, it demonstrates superior performance. The potential for future JIT optimizations strengthens the case for this method.

Any thoughts?

Reply:
The CheckedKeyword results also look decent. Considering exceptions should be rare, the extra 10 microseconds and 8 bytes might not be a big deal.

Reply:
I’d just go with the simple approach I detailed above as it lends the most capability for the JIT to optimize it.

We can revisit and optimize further if required later. Some of the easiest optimizations will be low hanging fruit such as improving `operator +`

Reply:
That is, I'd go with doing this:
```csharp
public static TimeSpan FromHours(int hours, long minutes = 0, long seconds = 0, long milliseconds = 0, long microseconds = 0)
{
    return FromHours(hours)
         + FromMinutes(minutes)
         + FromSeconds(seconds)
         + FromMilliseconds(milliseconds)
         + FromMicroseconds(microseconds);
}
```

The other patterns will end up with less capability for the JIT to optimize, especially for constant inputs, and that is overall less desirable.

Reply:
I opened https://github.com/dotnet/runtime/pull/94226 to fix up some of the cases like `operator +` so that the relevant inlining can occur.

Reply:
> I opened #94226 to fix up some of the cases like `operator +` so that the relevant inlining can occur.

Great, I already have the solution you wanted implemented, now I just have to write a few unit tests for it, although I'm not really certain what to test in specific. I guess I should test overflows and correct TimeSpans?